ndk/src/main/jni/bugsnag_ndk.c

```diff
@@ -0,0 +1,379 @@
+}
+
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_ndk_NativeBridge_deliverReportAtPath(
+    JNIEnv *env, jobject _this, jstring _report_path) {
+  const char *report_path = (*env)->GetStringUTFChars(env, _report_path, 0);
+  bugsnag_report *report =
+      bsg_deserialize_report_from_file((char *)report_path);
+
+  if (report != NULL) {
+    char *payload = bsg_serialize_report_to_json_string(report);
+    if (payload != NULL) {
+      jclass interface_class =
+          (*env)->FindClass(env, "com/bugsnag/android/NativeInterface");
+      jmethodID jdeliver_method =
+          (*env)->GetStaticMethodID(env, interface_class, "deliverReport",
+                                    "(Ljava/lang/String;Ljava/lang/String;)V");
+      jstring jpayload = (*env)->NewStringUTF(env, payload);
+      jstring jstage = (*env)->NewStringUTF(env, report->app.release_stage);
+      (*env)->CallStaticVoidMethod(env, interface_class, jdeliver_method,
+                                   jstage, jpayload);
+      (*env)->DeleteLocalRef(env, jpayload);
+      (*env)->DeleteLocalRef(env, jstage);
+      free(payload);
+    }
+    free(report);
+  }
+  remove(report_path);
+  (*env)->ReleaseStringUTFChars(env, _report_path, report_path);
+}
+
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_ndk_NativeBridge_addHandledEvent(JNIEnv *env,
+                                                          jobject _this) {
+  if (bsg_global_env == NULL)
+    return;
+  bsg_request_env_write_lock();
+  bsg_global_env->next_report.handled_events++;
+  bsg_release_env_write_lock();
+}
+
+JNIEXPORT void JNICALL Java_com_bugsnag_android_ndk_NativeBridge_startedSession(
+    JNIEnv *env, jstring session_id_, jlong start_date_) {
+  if (bsg_global_env == NULL || session_id_ == NULL)
+    return;
+  char *session_id = (char *)(*env)->GetStringUTFChars(env, session_id_, 0);
+  bsg_request_env_write_lock();
+  bugsnag_report_start_session(&bsg_global_env->next_report, session_id,
+                               (long)start_date_);
+  bsg_release_env_write_lock();
+  (*env)->ReleaseStringUTFChars(env, session_id_, session_id);
+}
+
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_ndk_NativeBridge_clearBreadcrumbs(JNIEnv *env,
+                                                           jobject _this) {
+  if (bsg_global_env == NULL)
+    return;
+  bsg_request_env_write_lock();
+  bugsnag_report_clear_breadcrumbs(&bsg_global_env->next_report);
+  bsg_release_env_write_lock();
+}
+
+JNIEXPORT void JNICALL Java_com_bugsnag_android_ndk_NativeBridge_addBreadcrumb(
+    JNIEnv *env, jobject _this, jstring name_, jstring crumb_type,
+    jstring timestamp_, jobject metadata) {
+  if (bsg_global_env == NULL)
+    return;
+  const char *name = (*env)->GetStringUTFChars(env, name_, 0);
+  const char *type = (*env)->GetStringUTFChars(env, crumb_type, 0);
+  const char *timestamp = (*env)->GetStringUTFChars(env, timestamp_, 0);
+  bugsnag_breadcrumb *crumb = calloc(1, sizeof(bugsnag_breadcrumb));
```

Comment:
Not the greatest; I may have possibly been dumping raw memory while debugging and its easier if its all in one place 😅

Reply:
Should we replace the `calloc` calls with `malloc` then? It jumped out to me as something to pay attention to as it was different which could confuse future travelers. It's also less performant right?

Reply:
I changed it just after reading the comment, and then changed it back because it introduced crashiness that I did not immediately understand. Will revisit later.

>  It's also less performant right?

At this minuscule scale, I doubt it.